### PR TITLE
[Feature] Add serveral differential privacy algorithms

### PR DESCRIPTION
**Pull request by DavdGao**
Main modifications: 
- Add functional mechanism for differential privacy, including:
    - a trainer named `fmtrainer` 
    - two models for linear regression and logistic regression in funcational mechanism
- Support gaussian and laplace DP mechanism by adding new optimizers
- Add unittest and scripts for DP mechanism

### CODE REVIEW

File: federatedscope/core/regularizer/l2_regularizer.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import torch
+
+REGULARIZER_NAME = "l2_regularizer"
+
+
+class L2Regularizer(Module):
+    """Returns the l2 norm of weight
+
+        Arguments:
+            p (int): The order of norm.
```
Comment (xieyxclack) [inline federatedscope/core/regularizer/l2_regularizer.py:13]: 'p=2'?

File: federatedscope/core/regularizer/l2_regularizer.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+        Returns:
+            Tensor: the norm of the given udpate.
+    """
+    def __init__(self):
+        super(L2Regularizer, self).__init__()
+
+    def forward(self, ctx):
+        l2_norm = 0.
+        for param in ctx.model.parameters():
```
Comment (xieyxclack) [inline federatedscope/core/regularizer/l2_regularizer.py:25]: An interesting discussion about whether l2 regularization should be applied to beta and gamma of BN layers: https://discuss.pytorch.org/t/weight-decay-in-the-optimizers-is-a-bad-idea-especially-with-batchnorm/16994.
IMO, we can keep the implementation same as that in `torch.norm`

**Review by xieyxclack — COMMENTED** [review]
LGTM, but I have not checked the implementation details of `Functional Mechanism: Regression Analysis under Differential Privacy`. And please refer to the inline comments for minor suggestions, thx!

File: federatedscope/core/regularizer/l2_regularizer.py (hunk 1 above)
Comment (DavdGao) [reply federatedscope/core/regularizer/l2_regularizer.py]: modified accordingly

File: federatedscope/core/regularizer/l2_regularizer.py (hunk 2 above)
Comment (DavdGao) [reply federatedscope/core/regularizer/l2_regularizer.py]: The current `torch.norm` will calculate the l2 norm for all parameters including bn weight and bias. 
For now, we use the parameter `skip_bn` to aovid calculating l2-norm for bn layer(with 'bn' in its name).

Comment (DavdGao) [reply federatedscope/core/regularizer/l2_regularizer.py]: Conflicts are solved

File: federatedscope/core/regularizer/l2_regularizer.py (hunk 1 above)
Comment (taxuexunshang) [reply federatedscope/core/regularizer/l2_regularizer.py]: In the funcational mechanism original paper, the noise addition process is performed only once for the objective function, while your implementation performs noise addition in each iteration, is this supported by the theory?
